### PR TITLE
tracing: Fix segfault when generated trace id disabled

### LINKED ISSUE #4792: tracing + 'generate_request_id: false' = segfault
**Bug**

*Title*: *tracing + 'generate_request_id: false' = segfault*

*Description*:

When upgrading from Envoy 1.6.0 to Envoy 1.8.,0 I found that Envoy 1.8.0 segfaults when running with tracing on and ``generate_request_id`` set to ``false``.

Below is the log on segfault. I generated this using a debug build of Envoy from the v1.8.0 tag. The binary from the official docker image segfaults too but the backtrace doesn't have code lines.

```
[2018-10-19 13:47:15.215][109992][info][main] source/server/server.cc:452] starting main dispatch loop
[2018-10-19 13:47:17.189][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:125] Caught Segmentation fault, suspect faulting address 0x0                                                                                                               
[2018-10-19 13:47:17.189][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:94] Backtrace thr<110001> obj<./envoy-bin-debug> (If unsymbolized, use tools/stack_decode.py):                                                                             
[2018-10-19 13:47:17.200][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #0 0x10e7a38 Envoy::Tracing::HttpTracerUtility::finalizeSpan()                                                                                            
[2018-10-19 13:47:17.211][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #1 0xfe2e3b Envoy::Http::ConnectionManagerImpl::ActiveStream::~ActiveStream()                                                                             
[2018-10-19 13:47:17.220][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #2 0xfe33f8 Envoy::Http::ConnectionManagerImpl::ActiveStream::~ActiveStream()                                                                             
[2018-10-19 13:47:17.229][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #3 0x6080fe std::default_delete<>::operator()()                                                                                                           
[2018-10-19 13:47:17.238][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #4 0xc2358b std::unique_ptr<>::reset()                                                                                                                    
[2018-10-19 13:47:17.246][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #5 0xc210ba Envoy::Event::DispatcherImpl::clearDeferredDeleteList()                                                                                       
[2018-10-19 13:47:17.254][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #6 0xc22d78 Envoy::Event::DispatcherImpl::DispatcherImpl()::$_0::operator()()                                                                             
[2018-10-19 13:47:17.262][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #7 0xc22c2c std::_Function_handler<>::_M_invoke()                                                                                                         
[2018-10-19 13:47:17.270][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #8 0x675b5d std::function<>::operator()()                                                                                                                 
[2018-10-19 13:47:17.278][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #9 0xc29ecc Envoy::Event::TimerImpl::TimerImpl()::$_0::operator()()                                                                                       
[2018-10-19 13:47:17.286][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #10 0xc29e98 Envoy::Event::TimerImpl::TimerImpl()::$_0::__invoke()                                                                                        
[2018-10-19 13:47:17.294][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #11 0x1341fd7 event_process_active_single_queue.isra.29                                                                                                   
[2018-10-19 13:47:17.301][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #12 0x13426ce event_process_active                                                                                                                        
[2018-10-19 13:47:17.309][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #13 0x1345457 event_base_loop                                                                                                                             
[2018-10-19 13:47:17.317][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #14 0xc2262c Envoy::Event::DispatcherImpl::run()                                                                                                          
[2018-10-19 13:47:17.325][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #15 0xc12615 Envoy::Server::WorkerImpl::threadRoutine()                                                                                                   
[2018-10-19 13:47:17.332][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #16 0xc135ee Envoy::Server::WorkerImpl::start()::$_2::operator()()                                                                                        
[2018-10-19 13:47:17.340][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #17 0xc1349c std::_Function_handler<>::_M_invoke()                                                                                                        
[2018-10-19 13:47:17.340][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #18 0x675b5d std::function<>::operator()()                                                                                                                
[2018-10-19 13:47:17.348][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #19 0x1353a7b Envoy::Thread::Thread::Thread()::$_0::operator()()                                                                                          
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #20 0x1353a54 Envoy::Thread::Thread::Thread()::$_0::__invoke()                                                                                            
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<110001> obj</lib64/libpthread.so.0>                                                                                                                               
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #21 0x7f12eb00e593 start_thread                                                                                                                           
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<110001> obj</lib64/libc.so.6>                                                                                                                                     
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<110001> #22 0x7f12ea9ade6e clone                                                                                                                                  
[2018-10-19 13:47:17.356][110001][critical][backtrace] bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:121] end backtrace thread 110001
```

*Repro steps*:

Example problem config:

```
admin:
  access_log_path: /tmp/envoy.log
  address:
    socket_address:
      address: 127.0.0.1
      port_value: 9902

tracing:
  http:
    name: envoy.zipkin
    config:
      collector_cluster: jaeger
      collector_endpoint: /api/v1/spans

static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 12802

    filter_chains:

      filters:
      - name: envoy.http_connection_manager
        config:
          server_name: sidecar
          generate_request_id: false
          stat_prefix: ingress_http
          tracing:
            operation_name: egress

          route_config:

            virtual_hosts:
              - name: wildcard
                domains:
                  - "*"
                routes:
                  - match:
                      prefix: "/"
                    route:
                      cluster: my_cluster

          http_filters:
          - name: envoy.router

  clusters:

    - name: my_cluster
      connect_timeout: 0.25s
      type: STATIC
      hosts:
        - socket_address:
            address: 127.0.0.1
            port_value: 1979

    - name: jaeger
      connect_timeout: 1s
      type: STATIC
      hosts:
        - socket_address:
            address: "127.0.0.1"
            port_value: "9411"
```

Running with this config and an http server listening on 1979, Envoy segfaults whenever I do an http requests on port 12802.

If I comment out the ``generate_request_id: false`` line then the segfault does not happen.

### PR DESCRIPTION
**Pull request by objectiser**
Signed-off-by: Gary Brown <gary@brownuk.com>

*Description*: Fixes #4792
*Risk Level*: low
*Testing*: Added unit test
*Docs Changes*: None required
*Release Notes*: None required
[Optional Fixes #Issue]
[Optional *Deprecated*:]

### CODE REVIEW

File: test/common/tracing/http_tracer_impl_test.cc [hunk 1]
```diff
@@ -139,6 +139,32 @@ TEST(HttpConnManFinalizerImpl, OriginalAndLongPath) {
   HttpTracerUtility::finalizeSpan(*span, &request_headers, stream_info, config);
 }
 
+TEST(HttpConnManFinalizerImpl, NoGeneratedId) {
+  const std::string path(300, 'a');
+  const std::string path_prefix = "http://";
+  const std::string expected_path(128, 'a');
+  std::unique_ptr<NiceMock<MockSpan>> span(new NiceMock<MockSpan>());
```
Comment (dnoe) [inline test/common/tracing/http_tracer_impl_test.cc]: Can this be a stack object as well?

Comment (objectiser) [reply test/common/tracing/http_tracer_impl_test.cc]: Sure - updated in other tests as well.

**Comment by dnoe** [comment]
Does the ipv6_tests failure occur when you run this locally? You can follow the directions here to do exactly what CI is doing for your testing: https://github.com/envoyproxy/envoy/blob/master/ci/README.md

**Comment by objectiser** [comment]
I tried running the `ci/do_circle_ci_ipv6_tests.sh` script locally, but get multiple failures - as I do when doing a normal build with `ENVOY_DOCKER_BUILD_DIR=~/build ./ci/run_envoy_docker.sh './ci/do_ci.sh bazel.dev'`. Probably something with my environment - but normally when doing a change I just check that there is no additional failures based on the changes I've made.

I had a look at the build failures for ipv6_tests, and couldn't see anything tracing related. Could you kick it off again, as I don't have the permissions - or alternatively I could push a simple change/rebase to trigger all tests again?

**Comment by objectiser** [comment]
Still failing - will take another look tomorrow.

**Pull request merged**